File: lab4/part1.m
clc; clear; close all;

% Генерация последовательности Голда
X = [0 1 1 0 1];
Y = [1 0 1 0 0];
L = 31;
g = zeros(1, L);

for i = 1:L
    g(i) = xor(X(end), Y(end));
    newX = xor(X(4), X(5));
    newY = xor(Y(2), Y(5));
    X = [newX X(1:end-1)];
    Y = [newY Y(1:end-1)];
end

disp('Последовательность Голда:');
fprintf('%d ', g);
fprintf('\n\n');

% Таблица автокорреляции 
for s = 0:L-1
    fprintf('%2d | ', s);
    
    % Вывод сдвинутой последовательности
    for i = 1:L
        idx = mod(i-1 + s, L) + 1;
        fprintf('%d ', g(idx));
    end
    
    % Вычисление автокорреляции
    matches = 0;
    mismatches = 0;
    for i = 1:L
        idx = mod(i-1 + s, L) + 1;
        if g(i) == g(idx)
            matches = matches + 1;
        else
            mismatches = mismatches + 1;
        end
    end
    
    diff = matches - mismatches;
    fprintf('| %d/31 (%.4f)\n', diff, diff/L);
end

% Преобразование для xcorr()
g_pm = 2*g - 1;

% График 1: Автокорреляция через xcorr()
[R_xc, lags_xc] = xcorr(g_pm, 'normalized');

figure('Name', 'Графики автокорреляции', 'Position', [100, 100, 1200, 500]);

subplot(1,3,1);
stem(lags_xc, R_xc, 'filled', 'MarkerSize', 4);
xlabel('Задержка (lag)');
ylabel('Автокорреляция');
title('xcorr() - нормализованная');
grid on;
xlim([-30, 30]);
hold on;
plot([0, 0], [-0.2, 1.1], 'r--', 'LineWidth', 0.5);

% График 2: Автокорреляция по формуле
R_formula = zeros(1, L);
for s = 0:L-1
    matches = 0;
    mismatches = 0;
    for i = 1:L
        idx = mod(i-1 + s, L) + 1;
        if g(i) == g(idx)
            matches = matches + 1;
        else
            mismatches = mismatches + 1;
        end
    end
    R_formula(s+1) = (matches - mismatches) / L;
end

subplot(1,3,2);
stem(0:L-1, R_formula, 'filled', 'MarkerSize', 4);
xlabel('Задержка (lag)');
ylabel('Автокорреляция');
title('По формуле из методички');
grid on;
xlim([0, 30]);
hold on;
plot([0, 30], [0, 0], 'k-', 'LineWidth', 0.5);

% График 3: Сравнение
subplot(1,3,3);
hold on;
stem(0:L-1, R_formula, 'b', 'filled', 'MarkerSize', 4, 'DisplayName', 'Формула');
[R_biased, lags] = xcorr(g_pm, 'biased');
stem(lags(lags>=0), R_biased(lags>=0), 'r', 'filled', 'MarkerSize', 3, 'DisplayName', 'xcorr(biased)');
xlabel('Задержка (lag)');
ylabel('Автокорреляция');
title('Сравнение методов');
legend('Location', 'best');
grid on;
xlim([0, 30]);

File: lab4/part1.c
#include <stdio.h>

int main() {
    int X[5] = {0,1,1,0,1};   // 13
    int Y[5] = {1,0,1,0,0};   // 20
    int g[31];

    for(int i = 0; i < 31; i++) {
        g[i] = X[4] ^ Y[4];  // выход = X4 xor Y4
        
        int newX = X[4] ^ X[3];  // X4 xor X5
        int newY = Y[4] ^ Y[1];  // Y2 xor Y5 
        
        for(int j = 4; j > 0; j--) {
            X[j] = X[j-1];
            Y[j] = Y[j-1];
        }
        X[0] = newX;
        Y[0] = newY;
    }

    
    for(int s = 0; s < 31; s++) {
        int matches = 0;
        int mismatches = 0;
        printf("%2d | ", s);

        for(int i = 0; i < 31; i++) {
            int b = g[(i + s) % 31];
            printf("%d ", b);
            if(g[i] == b) {
                matches++;
            } else {
                mismatches++;
            }
        }

        //  Rx(τ) = (число совпадений - число несовпадений) / N
        int diff = matches - mismatches;
        printf("| %d/31 (%.4f)\n", diff, (double)diff / 31);
    }
    
    return 0;
}